File: esdk-test-servers/go/handlers.go
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

// Operation handlers and the clientId registry. CreateClient builds the
// materials manager from the modeled config (delegating to the MPL), constructs
// the real ESDK client eagerly, and registers both under a fresh UUID. Encrypt
// and Decrypt resolve the registered client by clientId and forward to the real
// library, always through the materials manager.

package main

import (
	"context"
	"fmt"
	"reflect"
	"strings"
	"sync"

	mpl "github.com/aws/aws-cryptographic-material-providers-library/releases/go/mpl/awscryptographymaterialproviderssmithygenerated"
	mpltypes "github.com/aws/aws-cryptographic-material-providers-library/releases/go/mpl/awscryptographymaterialproviderssmithygeneratedtypes"
	esdkclient "github.com/aws/aws-encryption-sdk/releases/go/encryption-sdk/awscryptographyencryptionsdksmithygenerated"
	esdktypes "github.com/aws/aws-encryption-sdk/releases/go/encryption-sdk/awscryptographyencryptionsdksmithygeneratedtypes"
	"github.com/google/uuid"
)

// clientEntry is one registered ESDK client and the materials manager it
// encrypts and decrypts with.
type clientEntry struct {
	esdk *esdkclient.Client
	cmm  mpltypes.ICryptographicMaterialsManager
}

// appState is the shared server state: one MPL client for keyring/CMM
// construction and the thread-safe clientId -> clientEntry registry.
type appState struct {
	matProv *mpl.Client
	mu      sync.Mutex
	clients map[string]*clientEntry
}

func newAppState() (*appState, error) {
	matProv, err := mpl.NewClient(mpltypes.MaterialProvidersConfig{})
	if err != nil {
		return nil, err
	}
	return &appState{matProv: matProv, clients: make(map[string]*clientEntry)}, nil
}

// createClient eagerly builds the materials manager and the real ESDK client;
// any construction or validation failure is a GenericServerError.
func (s *appState) createClient(ctx context.Context, req *CreateClientRequest) (*CreateClientResponse, *serverError) {
	cmm, err := buildCMM(ctx, s.matProv, req.Config.CMM)
	if err != nil {
		return nil, generic("CreateClient failed to construct the ESDK client: %s", describeError(err))
	}
	policy := req.Config.CommitmentPolicy
	esdkConfig := esdktypes.AwsEncryptionSdkConfig{
		CommitmentPolicy:     &policy,
		MaxEncryptedDataKeys: req.Config.MaxEncryptedDataKeys,
	}
	// NewClient does not run the config's own validation, so run it here:
	// config failures (e.g. maxEncryptedDataKeys < 1) must surface at
	// CreateClient time.
	if err := esdkConfig.Validate(); err != nil {
		return nil, generic("CreateClient failed to construct the ESDK client: %s", describeError(err))
	}
	esdk, err := esdkclient.NewClient(esdkConfig)
	if err != nil {
		return nil, generic("CreateClient failed to construct the ESDK client: %s", describeError(err))
	}
	clientID := uuid.NewString()
	s.mu.Lock()
	s.clients[clientID] = &clientEntry{esdk: esdk, cmm: cmm}
	s.mu.Unlock()
	return &CreateClientResponse{ClientID: clientID}, nil
}

func (s *appState) encrypt(ctx context.Context, req *EncryptRequest) (*EncryptResponse, *serverError) {
	entry, serr := s.resolve(req.ClientID)
	if serr != nil {
		return nil, serr
	}
	out, err := entry.esdk.Encrypt(ctx, esdktypes.EncryptInput{
		Plaintext:         req.Plaintext,
		MaterialsManager:  entry.cmm,
		EncryptionContext: req.EncryptionContext,
		AlgorithmSuiteId:  req.AlgorithmSuiteID,
		FrameLength:       req.FrameLength,
	})
	if err != nil {
		return nil, esdkClientError(describeError(err))
	}
	return &EncryptResponse{Ciphertext: out.Ciphertext}, nil
}

func (s *appState) decrypt(ctx context.Context, req *DecryptRequest) (*DecryptResponse, *serverError) {
	entry, serr := s.resolve(req.ClientID)
	if serr != nil {
		return nil, serr
	}
	out, err := entry.esdk.Decrypt(ctx, esdktypes.DecryptInput{
		Ciphertext:        req.Ciphertext,
		MaterialsManager:  entry.cmm,
		EncryptionContext: req.EncryptionContext,
	})
	if err != nil {
		return nil, esdkClientError(describeError(err))
	}
	return &DecryptResponse{
		Plaintext:         out.Plaintext,
		EncryptionContext: out.EncryptionContext,
		AlgorithmSuiteID:  out.AlgorithmSuiteId,
	}, nil
}

// resolve looks up the registered client; a missing, empty, or unknown
// clientId is a GenericServerError.
func (s *appState) resolve(clientID string) (*clientEntry, *serverError) {
	if clientID == "" {
		return nil, generic("clientId must be non-empty")
	}
	s.mu.Lock()
	entry := s.clients[clientID]
	s.mu.Unlock()
	if entry == nil {
		return nil, generic("unknown clientId: %s", clientID)
	}
	return entry, nil
}

// describeError extracts the library error's message, flattening nested
// collected errors. The generated exception types share three field shapes
// across packages ({Message}, {ErrObject}, {ListOfErrors, Message}), and their
// Error() methods format a nil ErrorCodeOverride into the string, so read the
// fields reflectively instead.
func describeError(err error) string {
	v := reflect.ValueOf(err)
	if v.Kind() == reflect.Ptr && !v.IsNil() {
		v = v.Elem()
	}
	if v.Kind() != reflect.Struct {
		return err.Error()
	}
	message := ""
	described := false
	if f := v.FieldByName("Message"); f.IsValid() && f.Kind() == reflect.String && f.String() != "" {
		message = f.String()
		described = true
	}
	if f := v.FieldByName("ErrObject"); f.IsValid() && f.Kind() == reflect.Interface && !f.IsNil() {
		if inner, ok := f.Interface().(error); ok {
			message = describeError(inner)
		} else {
			message = fmt.Sprintf("%v", f.Interface())
		}
		described = true
	}
	if f := v.FieldByName("ListOfErrors"); f.IsValid() && f.Kind() == reflect.Slice && f.Len() > 0 {
		causes := make([]string, 0, f.Len())
		for i := 0; i < f.Len(); i++ {
			if inner, ok := f.Index(i).Interface().(error); ok {
				causes = append(causes, describeError(inner))
			}
		}
		encountered := fmt.Sprintf("[encountered: %s]", strings.Join(causes, "; "))
		if message != "" {
			message += " " + encountered
		} else {
			message = encountered
		}
		described = true
	}
	if !described {
		return err.Error()
	}
	return message
}

File: esdk-test-servers/rust/src/handlers.rs
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

//! Operation handlers and the `clientId` registry.
//!
//! `CreateClient` eagerly builds the materials manager from the modeled config
//! (delegating to the MPL) and the real ESDK client, registers both under a
//! fresh UUID, and returns the id. `Encrypt` and `Decrypt` resolve the entry by
//! `clientId` and drive the ESDK client through its materials manager.

use crate::error::{describe_esdk_error, ServerError};
use crate::keyring::build_cmm;
use crate::model::{
    CreateClientRequest, CreateClientResponse, DecryptRequest, DecryptResponse, EncryptRequest,
    EncryptResponse, EsdkAlgorithmSuiteId, EsdkCommitmentPolicy,
};
use aws_esdk::client::Client as EsdkClient;
use aws_esdk::material_providers::types::cryptographic_materials_manager::CryptographicMaterialsManagerRef;
use aws_esdk::material_providers::types::EsdkAlgorithmSuiteId as MplAlgorithmSuiteId;
use aws_esdk::material_providers::types::EsdkCommitmentPolicy as MplCommitmentPolicy;
use aws_esdk::AwsEncryptionSdkConfig;
use aws_smithy_types::Blob;
use std::collections::{BTreeMap, HashMap};
use std::sync::Mutex;

/// A registered client: the configured ESDK client and its materials manager.
#[derive(Clone)]
struct ClientEntry {
    esdk: EsdkClient,
    cmm: CryptographicMaterialsManagerRef,
}

/// Shared server state: the thread-safe `clientId -> ClientEntry` registry.
#[derive(Default)]
pub struct AppState {
    clients: Mutex<HashMap<String, ClientEntry>>,
}

fn map_commitment(policy: EsdkCommitmentPolicy) -> MplCommitmentPolicy {
    match policy {
        EsdkCommitmentPolicy::ForbidEncryptAllowDecrypt => {
            MplCommitmentPolicy::ForbidEncryptAllowDecrypt
        }
        EsdkCommitmentPolicy::RequireEncryptAllowDecrypt => {
            MplCommitmentPolicy::RequireEncryptAllowDecrypt
        }
        EsdkCommitmentPolicy::RequireEncryptRequireDecrypt => {
            MplCommitmentPolicy::RequireEncryptRequireDecrypt
        }
    }
}

fn to_mpl_suite(id: EsdkAlgorithmSuiteId) -> MplAlgorithmSuiteId {
    match id {
        EsdkAlgorithmSuiteId::Alg128GcmNoKdf => MplAlgorithmSuiteId::AlgAes128GcmIv12Tag16NoKdf,
        EsdkAlgorithmSuiteId::Alg192GcmNoKdf => MplAlgorithmSuiteId::AlgAes192GcmIv12Tag16NoKdf,
        EsdkAlgorithmSuiteId::Alg256GcmNoKdf => MplAlgorithmSuiteId::AlgAes256GcmIv12Tag16NoKdf,
        EsdkAlgorithmSuiteId::Alg128GcmHkdf => MplAlgorithmSuiteId::AlgAes128GcmIv12Tag16HkdfSha256,
        EsdkAlgorithmSuiteId::Alg192GcmHkdf => MplAlgorithmSuiteId::AlgAes192GcmIv12Tag16HkdfSha256,
        EsdkAlgorithmSuiteId::Alg256GcmHkdf => MplAlgorithmSuiteId::AlgAes256GcmIv12Tag16HkdfSha256,
        EsdkAlgorithmSuiteId::Alg128GcmHkdfEcdsa => {
            MplAlgorithmSuiteId::AlgAes128GcmIv12Tag16HkdfSha256EcdsaP256
        }
        EsdkAlgorithmSuiteId::Alg192GcmHkdfEcdsa => {
            MplAlgorithmSuiteId::AlgAes192GcmIv12Tag16HkdfSha384EcdsaP384
        }
        EsdkAlgorithmSuiteId::Alg256GcmHkdfEcdsa => {
            MplAlgorithmSuiteId::AlgAes256GcmIv12Tag16HkdfSha384EcdsaP384
        }
        EsdkAlgorithmSuiteId::Alg256GcmCommit => {
            MplAlgorithmSuiteId::AlgAes256GcmHkdfSha512CommitKey
        }
        EsdkAlgorithmSuiteId::Alg256GcmCommitEcdsa => {
            MplAlgorithmSuiteId::AlgAes256GcmHkdfSha512CommitKeyEcdsaP384
        }
    }
}

fn from_mpl_suite(id: MplAlgorithmSuiteId) -> EsdkAlgorithmSuiteId {
    match id {
        MplAlgorithmSuiteId::AlgAes128GcmIv12Tag16NoKdf => EsdkAlgorithmSuiteId::Alg128GcmNoKdf,
        MplAlgorithmSuiteId::AlgAes192GcmIv12Tag16NoKdf => EsdkAlgorithmSuiteId::Alg192GcmNoKdf,
        MplAlgorithmSuiteId::AlgAes256GcmIv12Tag16NoKdf => EsdkAlgorithmSuiteId::Alg256GcmNoKdf,
        MplAlgorithmSuiteId::AlgAes128GcmIv12Tag16HkdfSha256 => EsdkAlgorithmSuiteId::Alg128GcmHkdf,
        MplAlgorithmSuiteId::AlgAes192GcmIv12Tag16HkdfSha256 => EsdkAlgorithmSuiteId::Alg192GcmHkdf,
        MplAlgorithmSuiteId::AlgAes256GcmIv12Tag16HkdfSha256 => EsdkAlgorithmSuiteId::Alg256GcmHkdf,
        MplAlgorithmSuiteId::AlgAes128GcmIv12Tag16HkdfSha256EcdsaP256 => {
            EsdkAlgorithmSuiteId::Alg128GcmHkdfEcdsa
        }
        MplAlgorithmSuiteId::AlgAes192GcmIv12Tag16HkdfSha384EcdsaP384 => {
            EsdkAlgorithmSuiteId::Alg192GcmHkdfEcdsa
        }
        MplAlgorithmSuiteId::AlgAes256GcmIv12Tag16HkdfSha384EcdsaP384 => {
            EsdkAlgorithmSuiteId::Alg256GcmHkdfEcdsa
        }
        MplAlgorithmSuiteId::AlgAes256GcmHkdfSha512CommitKey => {
            EsdkAlgorithmSuiteId::Alg256GcmCommit
        }
        MplAlgorithmSuiteId::AlgAes256GcmHkdfSha512CommitKeyEcdsaP384 => {
            EsdkAlgorithmSuiteId::Alg256GcmCommitEcdsa
        }
    }
}

fn to_esdk_context(map: BTreeMap<String, String>) -> HashMap<String, String> {
    map.into_iter().collect()
}

impl AppState {
    pub fn new() -> Self {
        Self::default()
    }

    /// Eagerly build and register a client; returns its new id.
    pub async fn create_client(
        &self,
        req: CreateClientRequest,
    ) -> Result<CreateClientResponse, ServerError> {
        let cmm = build_cmm(&req.config.cmm).await?;
        let mut builder = AwsEncryptionSdkConfig::builder()
            .commitment_policy(map_commitment(req.config.commitment_policy));
        if let Some(max) = req.config.max_encrypted_data_keys {
            builder = builder.max_encrypted_data_keys(max);
        }
        let config = builder.build().map_err(|e| {
            ServerError::generic(format!("failed to construct the ESDK client: {e}"))
        })?;
        let esdk = EsdkClient::from_conf(config).map_err(|e| {
            ServerError::generic(format!(
                "failed to construct the ESDK client: {}",
                describe_esdk_error(&e)
            ))
        })?;
        let client_id = uuid::Uuid::new_v4().to_string();
        self.clients
            .lock()
            .expect("registry mutex")
            .insert(client_id.clone(), ClientEntry { esdk, cmm });
        Ok(CreateClientResponse { client_id })
    }

    pub async fn encrypt(&self, req: EncryptRequest) -> Result<EncryptResponse, ServerError> {
        let entry = self.entry(&req.client_id)?;
        let mut builder = entry
            .esdk
            .encrypt()
            .plaintext(Blob::new(req.plaintext))
            .materials_manager(entry.cmm);
        if let Some(context) = req.encryption_context {
            builder = builder.encryption_context(to_esdk_context(context));
        }
        if let Some(suite) = req.algorithm_suite_id {
            builder = builder.algorithm_suite_id(to_mpl_suite(suite));
        }
        if let Some(frame_length) = req.frame_length {
            builder = builder.frame_length(frame_length);
        }
        let output = builder
            .send()
            .await
            .map_err(|e| ServerError::esdk(describe_esdk_error(&e)))?;
        let ciphertext = output
            .ciphertext
            .map(Blob::into_inner)
            .ok_or_else(|| ServerError::generic("encrypt returned no ciphertext"))?;
        Ok(EncryptResponse { ciphertext })
    }

    pub async fn decrypt(&self, req: DecryptRequest) -> Result<DecryptResponse, ServerError> {
        let entry = self.entry(&req.client_id)?;
        let mut builder = entry
            .esdk
            .decrypt()
            .ciphertext(Blob::new(req.ciphertext))
            .materials_manager(entry.cmm);
        if let Some(context) = req.encryption_context {
            builder = builder.encryption_context(to_esdk_context(context));
        }
        let output = builder
            .send()
            .await
            .map_err(|e| ServerError::esdk(describe_esdk_error(&e)))?;
        let plaintext = output
            .plaintext
            .map(Blob::into_inner)
            .ok_or_else(|| ServerError::generic("decrypt returned no plaintext"))?;
        Ok(DecryptResponse {
            plaintext,
            encryption_context: output
                .encryption_context
                .filter(|map| !map.is_empty())
                .map(|map| map.into_iter().collect()),
            algorithm_suite_id: output.algorithm_suite_id.map(from_mpl_suite),
        })
    }

    /// Look up a registered client by id, cloning the shared handles.
    fn entry(&self, client_id: &str) -> Result<ClientEntry, ServerError> {
        if client_id.is_empty() {
            return Err(ServerError::generic("clientId must be non-empty"));
        }
        let clients = self.clients.lock().expect("registry mutex");
        clients
            .get(client_id)
            .cloned()
            .ok_or_else(|| ServerError::generic(format!("unknown clientId: {client_id}")))
    }
}
